balancer/rls/internal/child_policy.go

```diff
@@ -0,0 +1,216 @@
+//
+// Whenever a child policy wrapper changes its connectivity state, the LB policy
+// returns a new picker to the channel, since the channel may need to re-process
+// the picks for queued RPCs.
+//
+// It is not safe for concurrent access.
+type childPolicyWrapper struct {
+	bg     balancerGroup // BalancerGroup to which this child policy belongs.
+	logger *internalgrpclog.PrefixLogger
+
+	target        string                     // RLS target corresponding to this child policy.
+	targetField   string                     // Child policy config field to plug in the above target.
+	config        map[string]json.RawMessage // Child policy config.
+	builder       balancer.Builder           // Child policy builder to be passed to the balancer group.
+	parser        balancer.ConfigParser      // Child policy config parser for building the config.
+	resolverState resolver.State             // Resolver state as reported to the RLS LB policy.
+
+	refCnt           int                // Reference count.
+	state            balancer.State     // Balancer state reported by the child policy.
+	stateToAggregate connectivity.State // Child policy state reported for aggregation.
+}
+
+// childPolicyWrapperArgs is simply a collection of the arguments to be passed
+// to construct or update a child policy wrapper.
+type childPolicyWrapperArgs struct {
+	policyName    string
+	target        string
+	targetField   string
+	config        map[string]json.RawMessage
+	resolverState resolver.State
+	bg            balancerGroup
+}
+
+// newChildPolicyWrapper creates a new child policy wrapper for the given
+// arguments. The wrapper is initialized with no references and starts off in
+// CONNECTING state.
+//
+// It is important to note that the child policy is not added to the balancer
+// group at this point. It is added only when the reference to the child policy
+// is taken.
+func newChildPolicyWrapper(args childPolicyWrapperArgs) *childPolicyWrapper {
+	// Config parsing ensures that the child policy is registered.
+	builder := balancer.Get(args.policyName)
+	parser, _ := builder.(balancer.ConfigParser)
+
+	c := &childPolicyWrapper{
+		target:        args.target,
+		targetField:   args.targetField,
+		config:        args.config,
+		builder:       builder,
+		parser:        parser,
+		resolverState: args.resolverState,
+		bg:            args.bg,
+
+		refCnt: 0,
+		state: balancer.State{
+			ConnectivityState: connectivity.Connecting,
+			Picker:            base.NewErrPicker(balancer.ErrNoSubConnAvailable),
+		},
+		stateToAggregate: connectivity.Connecting,
+	}
+	c.logger = internalgrpclog.NewPrefixLogger(logger, fmt.Sprintf("[rls-child-policy-wrapper %s %p] ", c.target, c))
+	c.logger.Infof("Created")
+	return c
+}
+
+// handleNewConfigs handles updates to child policy configuration.
+func (c *childPolicyWrapper) handleNewConfigs(args childPolicyWrapperArgs) {
+	if c.builder.Name() != args.policyName {
```

Comment:
We're allowing the LB policy to be changed, but reusing the same wrapper?  Is that necessary?  Why not create a new wrapper when that happens instead?

Reply:
We should probably discuss offline since I'm not sure I'm following this.  It seems like if the entire cache is being rebuilt, then maybe invalidating the cache isn't a big deal?